### PR TITLE
Added price data by customer group id to product collection

### LINKED ISSUE #30083: [GraphQl bug] query Products in category is wrong price when set catalog price rule promotion for specific customer group
### Preconditions (*)

1. Magento 2.4.0

### Steps to reproduce (*)

1. Setup catalog price rule promotion with condition only customer login group for category A. The rule will discount 10% price for all products on this category.
2. Try to query graphql [products](https://devdocs.magento.com/guides/v2.4/graphql/queries/products.html) to get all product of category A with 2 situation:
- Not login
- Logged as customer
3. The data return as same with both. The price have not discounted with customer logged in yet

### Expected result (*)

1. The price of all products on category A will discount 10% with customer login request that query
2. 

### Actual result (*)

1. [Screenshots, logs or description]
2.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [x] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by andrewbess**
### Description (*)

This PR fixes the problem with wrong final prices for GraphQl query `products`

### Fixed Issues (if relevant)

1. Fixes magento/magento2#30083: [GraphQl bug] query Products in category is wrong price when set catalog price rule promotion for specific customer group

### Manual testing scenarios (*)

1. Create a new catalog price rule with discount for category for logged in customers
2. Generate customer token and write it to headers
3. Make a query `products` for this category as shown below
```
{
  products(filter: {category_id: {in: ["35"]}}) {
    items {
      sku
      name
      price_range {
        maximum_price {
          final_price {
            currency
            value
          }
          regular_price {
            currency
            value
          }
        }
        minimum_price {
          final_price {
            currency
            value
          }
          regular_price {
            currency
            value
          }
        }
      }
    }
  }
}
```
4. Check final prices and compare it with frontend

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by coderimus** [comment]
@magento run all tests

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+use Magento\CatalogGraphQl\Model\Resolver\Products\DataProvider\Product\CollectionProcessorInterface;
+use Magento\Customer\Api\CustomerRepositoryInterface;
+use Magento\Customer\Api\Data\GroupInterface;
+use Magento\Framework\Api\SearchCriteriaInterface;
+use Magento\GraphQl\Model\Query\ContextInterface;
+
+/**
+ * Add price data to product collection results
+ *
+ * @inheritdoc
```
Comment (coderimus) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php]: Please, remove the @inheridoc. It doesn't contain useful information and also SHOULD NOT be used according to the https://devdocs.magento.com/guides/v2.4/coding-standards/docblock-standard-general.html#inheritdoc

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    public function process(
+        Collection $collection,
+        SearchCriteriaInterface $searchCriteria,
+        array $attributeNames,
+        ContextInterface $context = null
+    ): Collection {
+        /**
+         * The $customerGroupId which is used to add price data to collection.
+         * Falls back to zero which represents the "Not Logged In" customer group.
+         */
```
Comment (coderimus) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php]: Please, remove this comment. The code your provided is self-described and easy to read 👍

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    /**
+     * @var CustomerRepositoryInterface
+     */
+    private $customerRepository;
+
+    /**
+     * @param CustomerRepositoryInterface $customerRepository
+     */
+    public function __construct(
+        CustomerRepositoryInterface $customerRepository
```
Comment (coderimus) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php]: Do we really need to make a new line here? Looks like this line is less then 120 symbols.

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    ): Collection {
+        /**
+         * The $customerGroupId which is used to add price data to collection.
+         * Falls back to zero which represents the "Not Logged In" customer group.
+         */
+        try {
+            if ($context
+                && $context->getExtensionAttributes()->getIsCustomer() === true
+            ) {
+                $customer = $this->customerRepository->getById(
```
Comment (coderimus) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php:56]: I think that this is a possible performance degradation point. Can we try to check the customer groupId by using the current customer session? `\Magento\Customer\Model\Session::getCustomerGroupId`

**Review by coderimus — CHANGES_REQUESTED** [review]
Hello @andrewbess ,
Thank you for the provided improvement. Please, check my notes below.

**Comment by andrewbess** [comment]
@magento run all tests

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php (hunk 1 above)
Comment (andrewbess) [reply app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php]: Hello @coderimus 
Thank you for your review.
I fixed it.
Please recheck.

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php (hunk 2 above)
Comment (andrewbess) [reply app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php]: Hello @coderimus 
Thank you for your review.
I fixed it.
Please recheck.

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php (hunk 4 above)
Comment (andrewbess) [reply app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php:56]: Hello @coderimus 
Thank you for your review.
I completely agree with you, but it seems to me that this is the problem of the context object of GraphQl and we should resolve it separately.

File: app/code/Magento/CatalogGraphQl/etc/module.xml [hunk 5]
```diff
@@ -15,6 +15,7 @@
             <module name="Magento_GraphQl"/>
             <module name="Magento_StoreGraphQl"/>
             <module name="Magento_EavGraphQl"/>
+            <module name="Magento_Customer"/>
```
Comment (rogyar) [inline app/code/Magento/CatalogGraphQl/etc/module.xml]: Please, do not forget to add the corresponding dependency to the `composer.json`

**Comment by andrewbess** [comment]
@magento run all tests

File: app/code/Magento/CatalogGraphQl/etc/module.xml (hunk 5 above)
Comment (andrewbess) [reply app/code/Magento/CatalogGraphQl/etc/module.xml]: Hello @rogyar 
Thank you for your review.
This PR doesn't require to change the sequence. But, it requires adding module `magento/module-customer` into `require` section of the `composer.json` file.
I fixed it. Please recheck.
Thank you in advance.

**Comment by andrewbess** [comment]
@magento run all tests

**Comment by andrewbess** [comment]
@magento run Functional Tests B2B

**Comment by andrewbess** [comment]
Hello @coderimus @rogyar 
Perhaps the problem isn't resolved fully.
It still doesn't work with configurable products and grouped products.
I am going to continue with it today evening.

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                    $context->getUserId()
+                );
+                $customerGroupId = (int) $customer->getGroupId();
+            } else {
+                $customerGroupId = GroupInterface::NOT_LOGGED_IN_ID;
+            }
+        } catch (Exception $e) {
```
Comment (kandy) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php:63]: Better to catch only here only known type of exception like \Magento\Framework\Exception\NoSuchEntityException

Comment (cpartica) [reply app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php:63]: I also think there's a setting in magento for default customer group
GroupInterface::NOT_LOGGED_IN_ID - is not really the default

You also need some tests to process this PR

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Product/PriceRange.php [hunk 7]
```diff
@@ -60,7 +60,6 @@ public function resolve(
 
         /** @var Product $product */
         $product = $value['model'];
-        $product->unsetData('minimal_price');
```
Comment (cpartica) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Product/PriceRange.php:63]: I am sure this had some other purpose and we probably broke something else here...

**Review by cpartica — CHANGES_REQUESTED** [review]

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php (hunk 6 above)
Comment (andrewbess) [reply app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Product/CollectionProcessor/PriceDataProcessor.php:63]: Hello @cpartica 
Here we require the group called "Not logged in", not the default group.

**Comment by rogyar** [comment]
Hi @andrewbess. Please, take a look at the recent comment above when you will have some time. 

Thank you!

**Comment by andrewbess** [comment]
The issue has been fixed with [this commit](https://github.com/magento/magento2/commit/e6e5c738eb47a982d90aef8e9a087cb3fd9bcacf) by the internal team.

**Comment by andrewbess** [comment]
Hello @rogyar 
I think issue #30080 should be closed.

**Pull request closed without merging**